One shot pem reader and writer

vcsjones opened the pull request:
Closes #29588

I thought I would give this one a shot.

/cc @bartonjs

vcsjones committed:
Find / TryFind PEM

vcsjones committed:
Implement GetEncodedSize.

vcsjones committed:
Added writing and complete doc-comments.

vcsjones commented:
> This serves as a reminder for when your PR is modifying a ref *.cs file and adding/modifying public APIs, to please make sure the API implementation in the src *.cs file is documented with triple slash comments, so the PR reviewers can sign off that change.

Well, I _tried_.

vcsjones commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,427 @@
+                Range postebLineRange = lineRange;
+                int postebLinePosition = preebLinePosition;
+
+                // in lax decoding a posteb may appear on the same line as the preeb.
```
This wasn't actually immediately clear to me from the RFC, if we even want to support single-line PEMs.

If not, this code can get a little simpler but I suspect someone will wish it "just worked".

vcsjones replied:
> Presumably, removing all newline processing (and ReadLine) and just using IndexOf to find preeb/posteb would actually make things simpler?

It still behaved as lax, but agree that it isn't necessary to handle line-by-line. I will address. It seems I may have misunderstood the whitespace requirement around the preeb / posteb for lax.

vcsjones committed:
Avoid an allocation for sensible label sizes.

vcsjones committed:
Stackalloc outside loop once.

vcsjones commented:
Looking at failures.

vcsjones committed:
Fix Windows line handling.

vcsjones commented:
OSX failures are #702.

ahsonkhan commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,435 @@
+                    continue;
+                }
+
+                int postebLength = Posteb.Length + label.Length + Ending.Length;
```
Could this overflow?

vcsjones replied:
I don't think so. `label`'s length is going to be based on a found label in the PEM, and we've already checked for the `"-----BEGIN "` and `"-----"`, so we know that the ROS (whose max length is an Int32) contains something of this length.

ahsonkhan commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,435 @@
+            for (int index = 1; index < data.Length; index++)
+            {
+                char c = data[index];
+                if (!IsLabelChar(c) && c != ' ' && c != '-')
```
Redundant `c != '-'` check?

vcsjones replied:
I don't think so. `IsLabelChar` says "Is this a valid character, except for "-"? The first character in allowed to be anything except a space or hyphen. The other characters may be any of those characters, including the hyphen or space. So it's a double negation. I'll try and make this easier to follow.

ahsonkhan commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,435 @@
+
+        private static bool IsValidLabel(ReadOnlySpan<char> data)
+        {
+            static bool IsLabelChar(char c) => c >= 0x21 && c <= 0x7e && c != '-';
```
Consider using "InRange" pattern:
https://source.dot.net/#System.Private.CoreLib/Char.cs,217

ahsonkhan commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,435 @@
+        }
+
+        /// <summary>
+        /// Finds the first PEM-encoded data.
```
```suggestion
        /// Attempts to find the first PEM-encoded data.
```

ahsonkhan commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemFields.cs:
```diff
@@ -0,0 +1,40 @@
+        /// <summary>
+        /// The size of the decoded base-64, in bytes.
+        /// </summary>
+        public int DecodedDataLength { get; }
```
I believe these properties need the `<value>` xml tag. @carlossanlop - is that correct?

carlossanlop replied:
Since Visual Studio doesn't add them automatically, we consider them optional in triple slash comments, but they are highly desirable, since we will eventually have to fill them out in dotnet-api-docs when these comments get ported there.

ahsonkhan replied:
> since we will eventually have to fill them out in dotnet-api-docs when these comments get ported there

So... because we have to eventually add them anyway, I think we should get in the habit of adding them in triple slash comments. This helps ensuring the docs/work happens as close to the PR/original API author. I don't see a benefit of leaving them out.

bartonjs replied:
For properties, long term, we probably want to either provide only value or only summary, and let the tool calculate the one from the other.  (Value -> summary is the easiest; but "just summary" is the most natural).

If a property summary starts with "Gets or sets " or "Gets ", value is almost always `summary.AsSpan(prefix, 1).ToUpperString() + summary.AsSpan(prefix+1)`.  If that turns out to not be the case (too many people correcting things in api-docs PRs) then we can do something else.

But given that it's almost redundant information (in English), I'm all for letting a tool do it, not a human.

ahsonkhan commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,435 @@
+            int precedingWhiteSpace = 0;
+            int trailingWhiteSpace = 0;
+            (int offset, int length) = content.GetOffsetAndLength(data.Length);
+            for (int index = offset; index < offset + length; index++)
```
I don't know how important perf is here, but I suspect this validation is quite slow (and doing the actual encoding using the Base64 APIs built on top of avx2/Ssse3 might actually be faster).

Alternatively, we could use some of the Base64 implementation as inspiration to make this method faster.

How large can `length` be?

vcsjones replied:
> but I suspect this validation is quite slow

Yeah, it occurred to me but when I looked around for inspiration such as at https://github.com/dotnet/runtime/blob/aac8f91cc5510b3d271d86eef616b6007b6e6631/src/libraries/System.Private.CoreLib/src/System/Convert.cs#L2820

I see it still doing character-by-character examination to account for various whitespace (albeit with unsafe tricks), just that the method has a looser definition of whitespace than the RFC.

Is there a better place to be looking here for this?

> How large can length be?

It could be very large, but the typical use-case for PEM is PKI and CMS. CMS can be several thousand characters.

lpereira replied:
Validation could be moved to `System.Buffers.Text.Base64`.  It is already vectorized, and contains a (private) decoding map that can be used to perform an efficient validation with vector instructions, with a scalar fallback.   The functions that decode using AVX2 and SSSE3 already check for invalid input before decoding, so it should be trivial.

ahsonkhan replied:
We would need to make that validation "whitespace" aware. The existing `S.B.T.Base64` APIs do not support ignoring whitespace characters.

ahsonkhan replied:
FWIW, something like this (which includes the work of actually decoding) is ~2x faster than doing the linear validation (for a ~1k buffer without any leading/trailing whitespace). I am not necessarily suggesting that this is the way to go (especially given the unnecessary work it does and is no where near "optimal"), but at the very least it avoids having to provide your own custom logic for the base64 validation. Ideally, this type of API is built-in to either the `Convert` or `Base64` classes.

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.19041
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.100-alpha1-015914
  [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.19.56303, CoreFX 5.0.19.56306), X64 RyuJIT
  Job-XEOKAL : .NET Core 5.0.0 (CoreCLR 42.42.42.42424, CoreFX 42.42.42.42424), X64 RyuJIT

PowerPlanMode=00000000-0000-0000-0000-000000000000  Toolchain=CoreRun  MaxIterationCount=10  
MinIterationCount=5  WarmupCount=3  

```
|           Method |     Mean |     Error |    StdDev |   Median |      Min |      Max | Ratio | Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------- |---------:|----------:|----------:|---------:|---------:|---------:|------:|------:|------:|------:|----------:|
| ValidateOriginal | 6.308 us | 0.1829 us | 0.1088 us | 6.316 us | 6.186 us | 6.548 us |  1.00 |     - |     - |     - |         - |
|      ValidateNew | 3.290 us | 0.0724 us | 0.0431 us | 3.302 us | 3.214 us | 3.352 us |  0.52 |     - |     - |     - |         - |


```C#
private static readonly char[] s_whitespace = new char[] { ' ', '\t', '\r', '\n' };
private static readonly char[] s_base64 = new char[]
{
    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
    'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h' ,'i', 'j',
    'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r' ,'s', 't',
    'u', 'v', 'w', 'x', 'y', 'z', 'A', 'B' ,'C', 'D',
    'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L' ,'M', 'N',
    'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V' ,'W', 'X',
    'Y', 'Z', '+', '/'
};

private bool IsValidBase64(out int decodedSize)
{
    ReadOnlySpan<char> input = _base64String.AsSpan();

    int startIndex = input.IndexOfAny(s_base64);

    if (startIndex == -1)
    {
        startIndex = 0;
    }

    input = input.Slice(startIndex);

    int finalIndex = input.LastIndexOfAny(s_whitespace);

    if (finalIndex == -1)
    {
        finalIndex = input.Length;
    }

    input = input.Slice(0, finalIndex);

    byte[] buffer = ArrayPool<byte>.Shared.Rent(Base64.GetMaxDecodedFromUtf8Length(input.Length));

    bool result = Convert.TryFromBase64Chars(input, buffer, out decodedSize);
    if (!result)
    {
        decodedSize = 0;
    }

    ArrayPool<byte>.Shared.Return(buffer, clearArray: true);

    return result;
}
```

ahsonkhan commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,435 @@
+        /// </param>
+        /// <returns>
+        /// A character array of the encoded PEM.
+        /// </returns>
```
Document exceptions.

ahsonkhan commented on src/libraries/System.Security.Cryptography.Encoding/tests/PemEncodingTests.cs:
```diff
@@ -0,0 +1,539 @@
+            Assert.Equal(3, fields.DecodedDataLength);
+        }
+
+
```
nit: remove extra line

vcsjones committed:
Code review feedback.

vcsjones committed:
Document value of PemFields.

vcsjones committed:
Document exception for Write and TryWrite.

vcsjones committed:
whitespace => white space

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/tests/PemEncodingTests.cs:
```diff
@@ -0,0 +1,548 @@
+        public static void TryFind_True_FindsPemAfterPemWithInvalidLabel()
+        {
+            string content = @"
+-----BEGIN ------
```
Probably worth a comment that this is an invalid label because there are too many hyphens after the space.

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/tests/PemEncodingTests.cs:
```diff
@@ -0,0 +1,548 @@
+        [Theory]
+        [InlineData("\tOOPS")]
+        [InlineData(" OOPS")]
+        [InlineData("-OOPS")]
```
```suggestion
        [InlineData("-OOPS")]
        [InlineData("OO  PS")]
```

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/tests/PemEncodingTests.cs:
```diff
@@ -0,0 +1,548 @@
+        }
+
+        [Fact]
+        public static void TryFind_False_PrecedingLinesAndSignificantCharsBeforePreeb()
```
I guess even under full lax, the character before the starting - of the pre-encapsulation boundary must not exist or be whitespace.

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/tests/PemEncodingTests.cs:
```diff
@@ -0,0 +1,548 @@
+        [Fact]
+        public static void TryFind_False_ContentOnPostEbLine()
+        {
+            string content = "-----BEGIN TEST-----\nZm9v\n-----END TEST-----boop";
```
I accept this interpretation, even under full lax.

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/tests/PemEncodingTests.cs:
```diff
@@ -0,0 +1,548 @@
+        [Fact]
+        public static void TryFind_False_IncompletePostEncapBoundary()
+        {
+            string content = "-----BEGIN TEST-----\nZm9v\n-----END TEST";
```
I don't think I saw a test, one way or the other, on mismatched labels between preeb and posteb.

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/tests/PemEncodingTests.cs:
```diff
@@ -0,0 +1,548 @@
+        public static void TryFind_False_NoPostEncapBoundary()
+        {
+            string content = "-----BEGIN TEST-----\nZm9v\n";
+            Assert.False(PemEncoding.TryFind(content, out _));
```
I don't know if we want subclassing, or double-duty; but it's probably good to have the tests always throw the same thing at both Find and TryFind, checking for the boolean/exception, and if TryFind succeeds that they give the same data.

With subclassing, it'd be something like

```C#
protected abstract PemFields Find(ReadOnlySpan<char> text);
protected abstract void FailToFind<TException>(ReadOnlySpan<char> text) where TException : Exception;

// (I have it returning the value, in case any multi-read tests wanted to compare two sets against each other)
private PemFields TestSuccess(string content, Range expectedLocation, Range expectedLabel, Range expectedBase64Data, int offset = 0)
{
    PemFields fields = Find(content.AsSpan(offset));
    // helper, as described before
}

public class PemEncodingFindTests_TryFind : PemEncodingFindTests
{
    protected override void FailToFind<TException>(ReadOnlySpan<char> text) where TException : Exception
    {
        Assert.False(PemEncoding.TryFind(text, out PemFields fields));
        Assert.True(fields.Location.Equals(default(Range)));
        ...
    }
}
```

lpereira commented:
Has this been fuzzed?

vcsjones committed:
Rewrite in terms without lines.

vcsjones committed:
Minor refactoring

vcsjones committed:
Add test for mismatched labels

ahsonkhan commented on src/libraries/System.Security.Cryptography.Encoding/src/Resources/Strings.resx:
```diff
@@ -117,6 +117,15 @@
     <value>Value was invalid.</value>
   </data>
+  <data name="Argument_PemEncoding_NoPemFound" xml:space="preserve">
+    <value>No PEM encoded data found.</value>
```
Should we explicitly state what PEM stands for in the exception message/API docs, or is it expected for the caller to have this domain knowledge?

Something like:
```suggestion
    <value>No Privacy-Enhanced Mail (PEM) encoded data found.</value>
```

bartonjs replied:
Nah, it'd be like if every JSON exception said "JavaScript Object Notation (JSON)".  The acronym explains the origin, but itself isn't relevant anymore.

ahsonkhan commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,457 @@
+            }
+
+            base64range = (offset + precedingWhiteSpace)..(offset + (length - trailingWhiteSpace));
+            decodedSize = (base64chars * 3) / 4;
```
I believe this give the incorrect results when the input span of char contains padding.
```C#
var r = new Random(42);
_bytes = new byte[1_000];
r.NextBytes(_bytes);
_base64String = Convert.ToBase64String(_bytes);
IsValidBase64(_base64String, out int decodedSize); // returns 1002, it should be 1000.
```

vcsjones replied:
Hm, thanks. I'll look at this, fix, and see overall what can be done about perf and correctness.

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,457 @@
+                    base64chars++;
+                }
+                else if (c == ' ' || c == '\n' || c == '\r' ||
+                         c == '\t' || c == '\v')
```
Convert.FromBase64String doesn't accept `\v`, so we probably only want to accept it (and form feed, which seems to be missing) as leading or trailing, but not infix.  Or just don't support VTab and FormFeed.

vcsjones replied:
> Or just don't support VTab and FormFeed.

That seems the most sensible.

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,457 @@
+                (c >= '0' && c <= '9') ||
+                (c >= 'A' && c <= 'Z') ||
+                (c >= 'a' && c <= 'z') ||
+                c == '+' || c == '/' || c == '=';
```
= is weird, since it's not a Base64 character; it's only allowed at the end.

vcsjones commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,457 @@
+                    continue;
+                }
+
+                (int preebOffset, _) = lineRange.GetOffsetAndLength(pemData.Length);
```
This is going to go away.

vcsjones commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,457 @@
+                Posteb.AsSpan().CopyTo(postebBuffer);
+                label.CopyTo(postebBuffer[Posteb.Length..]);
+                Ending.AsSpan().CopyTo(postebBuffer[(Posteb.Length + label.Length)..]);
+                ReadOnlySpan<char> posteb = postebBuffer[..postebLength];
```
> What is all the copying for?

At this point we know what the label is but we don't know where the full post encapsulation boundary is. To make "where is the full, correct, posteb" question easier to answer, this copies "-----END [LABEL]-----" in to the buffer (for reasonable labels will be on the stack) and we look for that whole string in one-go, rather than "-----END", "[LABEL]", "-----".

vcsjones commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,457 @@
+        /// <exception cref="ArgumentException">
+        /// The PEM is too large to possibly encode.
+        /// </exception>
+        public static char[] Write(ReadOnlySpan<char> label, ReadOnlySpan<byte> data)
```
Seems so? https://github.com/dotnet/runtime/issues/29588#issuecomment-582069797

vcsjones commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,457 @@
+            // and preeb lines new lines, assuming the base64 content is empty.
+            //     -----BEGIN {char * MaxLabelSize}-----\n
+            //     -----END {char * MaxLabelSize}-----
+            const int MaxLabelSize = 1_073_741_808;
```
>  Am I right?

Yes. I noted that Roslyn doesn't pre-compute this value for constant strings, so I just typed it out. Can do it with arithmetic instead, if desired.

vcsjones committed:
Fix label validation.

vcsjones committed:
Validate label when writing.

vcsjones committed:
Use clearer increment.

vcsjones committed:
Documentation feedback.

vcsjones commented:
@bartonjs @ahsonkhan @lpereira thanks y'all for the thorough feedback. This is going to take me some time to work through. To save review cycles, I would hold off on reviewing changes until I feel that it is in a reviewable state.

I would change the PR title to "WIP" for now, but I've been discouraged from doing so because it breaks email threading.

bartonjs commented:
Added `NO REVIEW`, just say something when it's ready

vcsjones committed:
Formatting cleanup.

vcsjones committed:
Remove line handling tests.

There is no more line-specific handling as part of moving to a better
interpretation of "lax".

danmosemsft commented:
Thanks for working on this, @vcsjones

vcsjones committed:
Some test refactoring.

ericsampson commented:
@vcsjones this would be wonderful to have. Will it be available as a package for netstandard2.0? 🙏

vcsjones committed:
WIP

vcsjones committed:
Merge remote-tracking branch 'ms/master' into one-shot-pem

vcsjones committed:
Fix tests and base64 location

vcsjones committed:
Use an offset helper to reduce IndexOf gymnastics.

vcsjones committed:
Rename

vcsjones committed:
Re-organize tests.

vcsjones committed:
Formatting

vcsjones committed:
Fixup decoding size calculation to not overflow an int

vcsjones committed:
Additional code review feedback.

vcsjones committed:
Remove duplicate tests.

vcsjones committed:
Add additional tests for base64 padding.

vcsjones committed:
Additional tests.

vcsjones commented:
@bartonjs 

I think this is ready for another round of review. Some comments:

1. I am not sure how best to address the performance suggestions of handling the base64. The summary of the discussion so far:
    - It's currently slower than the actual decoding implementation because  it's a straight forward implementation that looks at it character-by-character.
    - `System.Buffers.Text.Base64` is not suitable because it does not handle white space.
    - `Convert.*Base64*` APIs only do actual decoding. If we're worried about perf, we could decode in to a rented buffer.

    My level of comfort with performance-all-the-things isn't that high. My proposal would be to leave it as-is, and create a new API proposal somewhere that can use some of the existing decoding machinery to get the length with validation as well. I see utility of this as an API that is more generally available elsewhere. Alternatively, we could just use the existing decoding APIs and decode in to a scratch buffer. I expect most PEM encoded data to be a certificate or RSA-2048 PKCS1/8 key. That probably means decoding in to a stack buffer isn't worth implementing, so, pooled buffer. Decoding potentially private keys in to a pooled buffer makes me somewhat hesitant, but it looks like there are existing patterns to make that reasonable.

2. I got a little lost trying to follow the suggestions for doc-comments on `PemFields`, how `<value>` and `<summary>` relate. I think somewhere in there someone said "don't have both", so I stuck with `<summary>`.

vcsjones commented:
@ericsampson I don't think the intention was to ship this out of band. I don't know what's involved in that; or if there is a better place for these APIs to live that also ships out of band. @bartonjs?

ericsampson commented:
@vcsjones, thanks for considering if there is a way/place for it. There’s a lot of Core 2 apps that aren’t going away anytime soon :)

bartonjs commented:
Since it's not modifying existing types it's not impossible from a technical sense; but things get awkward if we release them as a NuGet package and also ship it as part of the framework for newer releases.

I can't say that it'll happen; but it's something that I'll bring up along with some other utility things in a similar bucket.

bartonjs commented on src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/AssertExtensions.cs:
```diff
@@ -413,9 +413,11 @@ public static void AtLeastOneEquals<T>(T expected1, T expected2, T value)
         }
 
-        public static void AssertThrows<E, T>(Span<T> span, AssertThrowsAction<T> action) where E : Exception
+        public static Exception AssertThrows<E, T>(Span<T> span, AssertThrowsAction<T> action) where E : Exception
```
Can't these return E as a stronger return type?

bartonjs commented on src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/AssertExtensions.cs:
```diff
@@ -438,6 +440,15 @@ public static void AtLeastOneEquals<T>(T expected1, T expected2, T value)
+            return exception;
+        }
+
+        public static void Throws<E, T>(string expectedParamName, ReadOnlySpan<T> span, AssertThrowsActionReadOnly<T> action)
```
The other `Throws(expectedParamName` methods return the exception; might as well go ahead and do it here for consistency (of type `E`).

And it probably makes sense to add the writable-span version for symmetry.

vcsjones committed:
Improve assert extensions.

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,518 @@
+                // The PostEB must either end at the end of the string, or
+                // have at least one white space character after it.
+                if (pemEndIndex < pemData.Length - 1 &&
+                    !char.IsWhiteSpace(pemData[pemEndIndex]))
```
@GrabYourPitchforks Do you have opinions on this "IsWhiteSpace" for determining the acceptability for end-of-payload?

The idea is `-----BEGIN FOO-----base64==----END FOO-----` is legal, `-----BEGIN FOO-----base64==----END FOO-----M` isn't (post-encapsulation-boundary might be part of some other text), but `-----BEGIN FOO-----base64==----END FOO----- M` is.  If the space is U+200A (hair space) is that enough to say we've hit the "end of the word" and thus the end of the post-EB?  Or would you prefer we stick to the RFC's `*W` production and limit it to `{ space, \t, \r, \n, \v, Form-feed }` or even our limited `*W` (`{ space, \t, \r, \n }`)?

GrabYourPitchforks replied:
Generally speaking we should never have a default behavior more _lax_ than the RFC (though we can be more _strict_ by default). This is the same philosophy we also followed with the JSON stack. So I'd suggest limiting this to RFC-blessed whitespace chars.

GrabYourPitchforks replied:
Ah, the _end_ of the payload. Nevermind then. No real opinions one way or another. :)

bartonjs replied:
@vcsjones Can you throw in a test that uses \u200A before and after to show that it satisfies the pre-preEB/post-postEB boundary constraints?  Alternatively, change the two IsWhiteSpaces to only allow space-tab-CR-LF and add a test that says we don't support the full set of IsWhiteSpace?

Since both @GrabYourPitchforks and I are iffy, it's your call on permissive or restrictive at the boundary.

vcsjones replied:
@bartonjs 
Since the base-64 decoding does not permit `/v` and `/f`, it made sense to me to limit the base-64 check to the more narrow definition of white space. It also makes sense to me to have a single definition of "white space" when decoding, so I changed the pre and post boundaries to match the same definition of white space as the base-64 validation.

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,518 @@
+                if (!TryCountBase64(pemData[contentRange],
+                                   out int base64start,
+                                   out int base64end,
+                                   out int decodedSize))
```
These look like they'd all fit on one line for the normal github aperture.  If you want/need the newlines, chop before the first argument and only indent 4 spaces.

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,518 @@
+
+                Range pemRange = preebIndex..pemEndIndex;
+                Range base64range = (contentStartIndex + base64start)..
+                                    (contentStartIndex + base64end);
```
```suggestion
                Range base64range = (contentStartIndex + base64start)..(contentStartIndex + base64end);
```

or

```suggestion
                Range base64range =
                    (contentStartIndex + base64start)..(contentStartIndex + base64end);
```

vcsjones committed:
Remove wrapping that is not required

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,518 @@
+                int size = PostEBPrefix.Length + label.Length + Ending.Length;
+                Debug.Assert(destination.Length >= size);
+                PostEBPrefix.AsSpan().CopyTo(destination);
+                label.CopyTo(destination[PostEBPrefix.Length..]);
```
I think that the left-defined-right-open calls look much nicer as explicitly saying `destination.Slice(PostEBPrefix.Length)`.

Maybe I'll feel different in a few years :)

vcsjones replied:
> Maybe I'll feel different in a few years :)

I kind of agree but felt weird mixing range-style slicing and literally calling `Slice`.

> I think that the left-defined-right-open calls look much nicer as explicitly saying 

Is this a stream of consciousness thought, or a "please change `[X..]` to `Slice(x)`??

bartonjs replied:
> Is this a stream of consciousness thought, or a "please change [X..] to Slice(x)??

An "I'd prefer" :smile:.

We don't have a style rule or strong convention; but some of us have a general leeriness of the syntax because `str[a..b]` is Substring, and `arr[a..b]` is effectively `Slice().ToArray()` -- the decision for range indexers is they return a collection of the same type, not a Span/Memory-equivalent projection.  For Span/ROSpan that means it's "zero"-cost slicing, but as syntax it's a bit of a burden on reviewers.

My personal filter is that I can accept the reviewer burden on "wait, are we a span type already?" for `[start..end]` (to avoid changing it to start/count); but for `[start..]` (`Slice(start)`) and `[..end]` (`Slice(0, end)`) the call to Slice reduces my cognitive load without adding increased load at changing parameter semantics.  (I'm also happy with changing `[start..end]` to `Slice(start, end - start)`, but it's more contextual)

vcsjones replied:
Seems reasonable. I removed the open-ended Range syntax in favor of explicit Slice. It's much to my preference to keep the [x..y] syntax for the reason you described.

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,518 @@
+
+                bool isSpaceOrHyphen = c == ' ' || c == '-';
+
+                if (!isSpaceOrHyphen || previousSpaceOrHyphen)
```
it's probably worth a comment that only one space-or-hyphen is allowed between label characters.  That logic can be re-derived from the code, but it's sort of "what is this doing, and why?".

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,518 @@
+                bool isSpaceOrHyphen = c == ' ' || c == '-';
+
+                if (!isSpaceOrHyphen || previousSpaceOrHyphen)
+                    return false;
```
Unbraced statements aren't allowed within braced statements.  (Which has a net result of: they're only allowed in the outer method scope)

```suggestion
                {
                    return false;
                }
```

vcsjones replied:
@bartonjs okay, I made this mistake a number of times. I will address all of them.

vcsjones committed:
Add comment explaining label composition.

vcsjones committed:
Brace for impact.

vcsjones committed:
Remove open-ended ranges and use Slice.

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs:
```diff
@@ -0,0 +1,532 @@
+                previousSpaceOrHyphen = true;
+            }
+
+            return true;
```
A space-or-hyphen has to be followed by another labelchar.  So isn't this `return !previousSpaceOrHyphen;`?

I think right now it'll say "HELLO-" is a valid label. (I don't see a test confirming one way or another)

vcsjones replied:
You're correct. I misread the RFC.

vcsjones committed:
Correct IsValidLabel.

Last character must be a labelchar.

vcsjones committed:
Test additions.

bartonjs commented on src/libraries/System.Security.Cryptography.Encoding/tests/PemEncodingFindTests.cs:
```diff
@@ -0,0 +1,381 @@
+        [InlineData("H-E-L-L-O")]
+        [InlineData("HEL-LO")]
+        [InlineData("H")]
+        public void TryFind_True_LabelsWithHyphenSpace(string label)
```
"H" has no hyphens or spaces, so ideally either it moves to a different method, or the name just gets updated here.

It also looks like the name says TryFind but the body uses Find.

vcsjones committed:
Fix test names

vcsjones committed:
More specific definition of white space.
